File: src/StreamLZ/Common/FrameFormat.cs
// FrameFormat.cs — StreamLZ frame format constants, header types, and serialization.

using System.Buffers.Binary;
using System.Numerics;
using System.Runtime.CompilerServices;

namespace StreamLZ.Common;

/// <summary>
/// Constants for the StreamLZ frame format (version 1).
/// </summary>
public static class FrameConstants
{
    /// <summary>Magic number identifying a StreamLZ frame: "SLZ1" (0x314C5A53 little-endian).</summary>
    public const uint MagicNumber = 0x53_4C_5A_31; // 'S','L','Z','1' in memory order

    /// <summary>Current frame format version.</summary>
    public const byte Version = 1;

    /// <summary>End-of-stream marker: a block with compressed size = 0.</summary>
    public const uint EndMark = 0;

    /// <summary>High bit in block compressed size indicating uncompressed (stored) data.</summary>
    public const uint BlockUncompressedFlag = 0x80000000;

    /// <summary>Minimum frame header size (no optional fields): magic + version + flags + codec + level + blockSize + reserved = 10 bytes.</summary>
    public const int MinHeaderSize = 10;

    /// <summary>Maximum frame header size (all optional fields): min + contentSize(8) + dictId(4) = 22 bytes.</summary>
    public const int MaxHeaderSize = 22;

    /// <summary>Default block size for framed compression (256 KB, matching the chunk size).</summary>
    public const int DefaultBlockSize = StreamLZConstants.ChunkSize;

    /// <summary>Minimum supported block size (64 KB).</summary>
    public const int MinBlockSize = 0x10000;

    /// <summary>Maximum supported block size for serial path (4 MB).</summary>
    public const int MaxBlockSize = 0x400000;

    /// <summary>Maximum decompressed block size accepted from a stream (512 MB).
    /// Guards against malicious streams claiming enormous block sizes to force allocation.</summary>
    public const int MaxDecompressedBlockSize = 512 * 1024 * 1024;

    /// <summary>Default sliding window size (128 MB).</summary>
    public const int DefaultWindowSize = 128 * 1024 * 1024;

    /// <summary>Maximum sliding window size (1 GB, matching MaxDictionarySize).</summary>
    public const int MaxWindowSize = StreamLZConstants.MaxDictionarySize;
}

/// <summary>
/// Flags stored in the frame header controlling optional features.
/// Underlying type is <c>byte</c> because it maps directly to a single header byte on the wire.
/// </summary>
[Flags]
[System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1028", Justification = "Byte maps directly to wire format")]
[System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1711", Justification = "Flags suffix is descriptive for a [Flags] enum")]
public enum FrameFlags : byte
{
    /// <summary>No optional features.</summary>
    None = 0,

    /// <summary>8-byte content size is present after the fixed header fields.</summary>
    ContentSizePresent = 1 << 0,

    /// <summary>XXH32 content checksum is appended after the end mark.</summary>
    ContentChecksum = 1 << 1,

    /// <summary>XXH32 block checksum is appended after each compressed block.</summary>
    BlockChecksums = 1 << 2,

    /// <summary>4-byte dictionary ID is present after content size (if any).</summary>
    DictionaryIdPresent = 1 << 3,
}

/// <summary>
/// Parsed frame header describing the compressed stream.
/// </summary>
public readonly struct FrameHeader : IEquatable<FrameHeader>
{
    /// <summary>Frame format version (currently 1).</summary>
    public byte Version { get; init; }

    /// <summary>Feature flags.</summary>
    public FrameFlags Flags { get; init; }

    /// <summary>Codec used for compression (0 = High, 1 = Fast).</summary>
    public byte Codec { get; init; }

    /// <summary>Compression level (0-9).</summary>
    public byte Level { get; init; }

    /// <summary>Block size in bytes (decoded from log2 representation).</summary>
    public int BlockSize { get; init; }

    /// <summary>Original uncompressed content size, or -1 if not present.</summary>
    public long ContentSize { get; init; }

    /// <summary>Dictionary ID, or 0 if not present.</summary>
    public uint DictionaryId { get; init; }

    /// <summary>Total number of header bytes (for skipping past the header on read).</summary>
    public int HeaderSize { get; init; }

    /// <inheritdoc/>
    public bool Equals(FrameHeader other) =>
        Version == other.Version && Flags == other.Flags && Codec == other.Codec &&
        Level == other.Level && BlockSize == other.BlockSize && ContentSize == other.ContentSize &&
        DictionaryId == other.DictionaryId && HeaderSize == other.HeaderSize;

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is FrameHeader other && Equals(other);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(Version, Flags, Codec, Level, BlockSize, ContentSize, DictionaryId, HeaderSize);

    /// <summary>Equality operator.</summary>
    public static bool operator ==(FrameHeader left, FrameHeader right) => left.Equals(right);

    /// <summary>Inequality operator.</summary>
    public static bool operator !=(FrameHeader left, FrameHeader right) => !left.Equals(right);
}

/// <summary>
/// Reads and writes StreamLZ frame headers.
/// </summary>
public static class FrameSerializer
{
    /// <summary>
    /// Writes a frame header to the destination buffer.
    /// </summary>
    /// <param name="destination">Buffer to write the header to (must be at least <see cref="FrameConstants.MaxHeaderSize"/> bytes).</param>
    /// <param name="codec">Codec identifier (0 = High, 1 = Fast).</param>
    /// <param name="level">Compression level (0-9).</param>
    /// <param name="blockSize">Block size in bytes (must be a power of 2 between 64KB and 4MB).</param>
    /// <param name="contentSize">Original content size, or -1 if unknown.</param>
    /// <param name="useContentChecksum">Whether to write a content checksum after the last block.</param>
    /// <param name="useBlockChecksums">Whether to write per-block checksums.</param>
    /// <returns>Number of header bytes written.</returns>
    public static int WriteHeader(Span<byte> destination, int codec, int level,
        int blockSize = FrameConstants.DefaultBlockSize, long contentSize = -1,
        bool useContentChecksum = false, bool useBlockChecksums = false)
    {
        FrameFlags flags = FrameFlags.None;
        if (contentSize >= 0) flags |= FrameFlags.ContentSizePresent;
        if (useContentChecksum) flags |= FrameFlags.ContentChecksum;
        if (useBlockChecksums) flags |= FrameFlags.BlockChecksums;

        if (!BitOperations.IsPow2(blockSize))
            throw new ArgumentException($"Block size must be a power of 2, got {blockSize}.", nameof(blockSize));

        int blockSizeLog2 = BitOperationsLog2(blockSize) - BitOperationsLog2(FrameConstants.MinBlockSize);

        int pos = 0;
        BinaryPrimitives.WriteUInt32LittleEndian(destination[pos..], FrameConstants.MagicNumber);
        pos += 4;
        destination[pos++] = FrameConstants.Version;
        destination[pos++] = (byte)flags;
        destination[pos++] = (byte)codec;
        destination[pos++] = (byte)level;
        destination[pos++] = (byte)blockSizeLog2;
        destination[pos++] = 0; // reserved

        if ((flags & FrameFlags.ContentSizePresent) != 0)
        {
            BinaryPrimitives.WriteInt64LittleEndian(destination[pos..], contentSize);
            pos += 8;
        }

        return pos;
    }

    /// <summary>
    /// Reads and parses a frame header from the source buffer.
    /// </summary>
    /// <param name="source">Buffer containing the frame header.</param>
    /// <param name="header">Receives the parsed header.</param>
    /// <returns>True if the header was parsed successfully, false if the magic number doesn't match or the data is too short.</returns>
    public static bool TryReadHeader(ReadOnlySpan<byte> source, out FrameHeader header)
    {
        header = default;

        if (source.Length < FrameConstants.MinHeaderSize)
            return false;

        uint magic = BinaryPrimitives.ReadUInt32LittleEndian(source);
        if (magic != FrameConstants.MagicNumber)
            return false;

        int pos = 4;
        byte version = source[pos++];
        if (version != FrameConstants.Version)
            return false;

        var flags = (FrameFlags)source[pos++];
        byte codec = source[pos++];
        byte level = source[pos++];

        int blockSizeLog2 = source[pos++] + BitOperationsLog2(FrameConstants.MinBlockSize);
        if (blockSizeLog2 < BitOperationsLog2(FrameConstants.MinBlockSize) || blockSizeLog2 > BitOperationsLog2(FrameConstants.MaxBlockSize))
            return false;
        int blockSize = 1 << blockSizeLog2;
        pos++; // reserved

        long contentSize = -1;
        if ((flags & FrameFlags.ContentSizePresent) != 0)
        {
            if (source.Length < pos + 8)
                return false;
            contentSize = BinaryPrimitives.ReadInt64LittleEndian(source[pos..]);
            pos += 8;
        }

        uint dictionaryId = 0;
        if ((flags & FrameFlags.DictionaryIdPresent) != 0)
        {
            if (source.Length < pos + 4)
                return false;
            dictionaryId = BinaryPrimitives.ReadUInt32LittleEndian(source[pos..]);
            pos += 4;
        }

        header = new FrameHeader
        {
            Version = version,
            Flags = flags,
            Codec = codec,
            Level = level,
            BlockSize = blockSize,
            ContentSize = contentSize,
            DictionaryId = dictionaryId,
            HeaderSize = pos,
        };
        return true;
    }

    /// <summary>
    /// Writes an end-of-stream mark (4 zero bytes).
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void WriteEndMark(Span<byte> destination)
    {
        BinaryPrimitives.WriteUInt32LittleEndian(destination, FrameConstants.EndMark);
    }

    /// <summary>
    /// Writes a block header: 4-byte LE compressed size + 4-byte LE decompressed size.
    /// High bit of compressed size indicates uncompressed (stored) data.
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void WriteBlockHeader(Span<byte> destination, int compressedSize, int decompressedSize, bool isUncompressed)
    {
        uint value = (uint)compressedSize;
        if (isUncompressed) value |= FrameConstants.BlockUncompressedFlag;
        BinaryPrimitives.WriteUInt32LittleEndian(destination, value);
        BinaryPrimitives.WriteInt32LittleEndian(destination[4..], decompressedSize);
    }

    /// <summary>
    /// Reads a block header: returns compressed size, decompressed size, and whether the block is uncompressed.
    /// A compressed size of 0 indicates end-of-stream.
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool TryReadBlockHeader(ReadOnlySpan<byte> source, out int compressedSize, out int decompressedSize, out bool isUncompressed)
    {
        compressedSize = 0;
        decompressedSize = 0;
        isUncompressed = false;
        if (source.Length < 8)
            return false;
        uint value = BinaryPrimitives.ReadUInt32LittleEndian(source);
        if (value == FrameConstants.EndMark)
            return true; // end mark: compressedSize = 0
        isUncompressed = (value & FrameConstants.BlockUncompressedFlag) != 0;
        compressedSize = (int)(value & ~FrameConstants.BlockUncompressedFlag);
        decompressedSize = BinaryPrimitives.ReadInt32LittleEndian(source[4..]);
        return compressedSize > 0 && decompressedSize > 0;
    }

    private static int BitOperationsLog2(int value)
    {
        return System.Numerics.BitOperations.Log2((uint)value);
    }
}

File: src/StreamLZ_zig/src/format/frame_format.zig
//! Port of src/StreamLZ/Common/FrameFormat.cs — SLZ1 frame format.
//!
//! Wire layout (little-endian):
//!   [4] magic = 0x534C5A31 ('SLZ1' byte order S,L,Z,1)
//!   [1] version = 1
//!   [1] flags
//!   [1] codec (0=High, 1=Fast, 2=Turbo)
//!   [1] level (internal: High L5/L7/L9 or Fast L1/L2/L3/L5/L6)
//!   [1] block_size_log2 (offset from log2(min_block_size) = 16)
//!   [1] reserved
//!   [8] content_size (LE i64, present iff flags.ContentSize)
//!   [4] dictionary_id (LE u32, present iff flags.DictionaryId)
//!
//! Then blocks:
//!   [4] compressed_size (LE u32; high bit = uncompressed; == 0 terminates)
//!   [4] decompressed_size (LE u32)
//!   [compressed_size] block payload
//!
//! After end mark (4 zeros):
//!   [4] XXH32 content checksum (LE u32, if flags.ContentChecksum)

const std = @import("std");
const constants = @import("streamlz_constants.zig");

pub const magic: u32 = 0x534C5A31;
pub const version: u8 = 1;
pub const end_mark: u32 = 0;

pub const block_uncompressed_flag: u32 = 0x80000000;

pub const min_header_size: usize = 10;
pub const max_header_size: usize = 22;

pub const default_block_size: usize = constants.chunk_size;
pub const min_block_size: usize = 0x10000; // 64 KB
pub const max_block_size: usize = 0x400000; // 4 MB
pub const max_decompressed_block_size: usize = 512 * 1024 * 1024;

pub const default_window_size: usize = 128 * 1024 * 1024;
pub const max_window_size: usize = constants.max_dictionary_size;

pub const FrameFlags = packed struct(u8) {
    content_size_present: bool = false,
    content_checksum: bool = false,
    block_checksums: bool = false,
    dictionary_id_present: bool = false,
    _reserved: u4 = 0,
};

pub const Codec = enum(u8) {
    high = 0,
    fast = 1,
    turbo = 2,
    _,

    pub fn name(self: Codec) []const u8 {
        return switch (self) {
            .high => "High",
            .fast => "Fast",
            .turbo => "Turbo",
            else => "Unknown",
        };
    }
};

pub const FrameHeader = struct {
    version: u8,
    flags: FrameFlags,
    codec: Codec,
    level: u8,
    block_size: u32,
    content_size: ?u64,
    dictionary_id: ?u32,
    header_size: usize,
};

pub const ParseError = error{
    BadMagic,
    UnsupportedVersion,
    BadBlockSize,
    Truncated,
};

/// Reads and parses a frame header from `src`. Returns the parsed header
/// and the number of bytes consumed (header_size).
pub fn parseHeader(src: []const u8) ParseError!FrameHeader {
    if (src.len < min_header_size) return error.Truncated;

    const got_magic = std.mem.readInt(u32, src[0..4], .little);
    if (got_magic != magic) return error.BadMagic;

    var pos: usize = 4;
    const ver = src[pos];
    pos += 1;
    if (ver != version) return error.UnsupportedVersion;

    const raw_flags: FrameFlags = @bitCast(src[pos]);
    pos += 1;
    const codec: Codec = @enumFromInt(src[pos]);
    pos += 1;
    const lvl = src[pos];
    pos += 1;

    const min_log2 = std.math.log2_int(usize, min_block_size);
    const max_log2 = std.math.log2_int(usize, max_block_size);
    const block_size_log2_encoded = src[pos];
    pos += 1;
    const block_size_log2 = @as(u8, @intCast(min_log2)) + block_size_log2_encoded;
    if (block_size_log2 < min_log2 or block_size_log2 > max_log2) return error.BadBlockSize;
    const block_size: u32 = @as(u32, 1) << @intCast(block_size_log2);

    pos += 1; // reserved

    var content_size: ?u64 = null;
    if (raw_flags.content_size_present) {
        if (src.len < pos + 8) return error.Truncated;
        const cs_raw = std.mem.readInt(i64, src[pos..][0..8], .little);
        content_size = if (cs_raw >= 0) @intCast(cs_raw) else null;
        pos += 8;
    }

    var dict_id: ?u32 = null;
    if (raw_flags.dictionary_id_present) {
        if (src.len < pos + 4) return error.Truncated;
        dict_id = std.mem.readInt(u32, src[pos..][0..4], .little);
        pos += 4;
    }

    return .{
        .version = ver,
        .flags = raw_flags,
        .codec = codec,
        .level = lvl,
        .block_size = block_size,
        .content_size = content_size,
        .dictionary_id = dict_id,
        .header_size = pos,
    };
}

pub const WriteHeaderOptions = struct {
    codec: Codec,
    level: u8,
    block_size: u32 = default_block_size,
    content_size: ?u64 = null,
    content_checksum: bool = false,
    block_checksums: bool = false,
    dictionary_id: ?u32 = null,
};

/// Writes a frame header to `dst` and returns the number of bytes written.
/// Caller must ensure `dst.len >= max_header_size`.
pub fn writeHeader(dst: []u8, opts: WriteHeaderOptions) !usize {
    if (opts.level < 1 or opts.level > 9) return error.BadLevel;
    if (opts.block_size < min_block_size or opts.block_size > max_block_size) return error.BadBlockSize;
    if (!std.math.isPowerOfTwo(opts.block_size)) return error.BadBlockSize;

    const flags: FrameFlags = .{
        .content_size_present = opts.content_size != null,
        .content_checksum = opts.content_checksum,
        .block_checksums = opts.block_checksums,
        .dictionary_id_present = opts.dictionary_id != null,
    };

    var pos: usize = 0;
    std.mem.writeInt(u32, dst[pos..][0..4], magic, .little);
    pos += 4;
    dst[pos] = version;
    pos += 1;
    dst[pos] = @bitCast(flags);
    pos += 1;
    dst[pos] = @intFromEnum(opts.codec);
    pos += 1;
    dst[pos] = opts.level;
    pos += 1;

    const min_log2: u8 = @intCast(std.math.log2_int(usize, min_block_size));
    const this_log2: u8 = @intCast(std.math.log2_int(u32, opts.block_size));
    dst[pos] = this_log2 - min_log2;
    pos += 1;
    dst[pos] = 0; // reserved
    pos += 1;

    if (opts.content_size) |cs| {
        std.mem.writeInt(i64, dst[pos..][0..8], @intCast(cs), .little);
        pos += 8;
    }
    if (opts.dictionary_id) |id| {
        std.mem.writeInt(u32, dst[pos..][0..4], id, .little);
        pos += 4;
    }
    return pos;
}

pub const BlockHeader = struct {
    compressed_size: u32,
    decompressed_size: u32,
    uncompressed: bool,

    /// True when this is the end-of-stream sentinel (compressed_size == 0).
    pub fn isEndMark(self: BlockHeader) bool {
        return self.compressed_size == 0;
    }
};

pub fn parseBlockHeader(src: []const u8) ParseError!BlockHeader {
    // End mark is just 4 zero bytes — no decompressed size field.
    if (src.len < 4) return error.Truncated;
    const raw = std.mem.readInt(u32, src[0..4], .little);
    if (raw == end_mark) {
        return .{ .compressed_size = 0, .decompressed_size = 0, .uncompressed = false };
    }
    if (src.len < 8) return error.Truncated;
    const decompressed = std.mem.readInt(u32, src[4..8], .little);
    return .{
        .compressed_size = raw & ~block_uncompressed_flag,
        .decompressed_size = decompressed,
        .uncompressed = (raw & block_uncompressed_flag) != 0,
    };
}

pub fn writeBlockHeader(dst: []u8, hdr: BlockHeader) void {
    var raw: u32 = hdr.compressed_size;
    if (hdr.uncompressed) raw |= block_uncompressed_flag;
    std.mem.writeInt(u32, dst[0..4], raw, .little);
    std.mem.writeInt(u32, dst[4..8], hdr.decompressed_size, .little);
}

pub fn writeEndMark(dst: []u8) void {
    std.mem.writeInt(u32, dst[0..4], end_mark, .little);
}

// ────────────────────────────────────────────────────────────
//  Tests
// ────────────────────────────────────────────────────────────

const testing = std.testing;

test "parseHeader rejects non-SLZ1 magic" {
    const bogus = [_]u8{ 'N', 'O', 'P', 'E', 1, 0, 0, 1, 2, 0 };
    try testing.expectError(error.BadMagic, parseHeader(&bogus));
}

test "parseHeader rejects truncated input" {
    const tiny = [_]u8{ 0x31, 0x5A, 0x4C, 0x53 };
    try testing.expectError(error.Truncated, parseHeader(&tiny));
}

test "writeHeader / parseHeader roundtrip, minimal flags" {
    var buf: [max_header_size]u8 = undefined;
    const n = try writeHeader(&buf, .{
        .codec = .fast,
        .level = 1,
        .block_size = default_block_size,
    });
    const hdr = try parseHeader(buf[0..n]);
    try testing.expectEqual(@as(u8, 1), hdr.version);
    try testing.expectEqual(Codec.fast, hdr.codec);
    try testing.expectEqual(@as(u8, 1), hdr.level);
    try testing.expectEqual(@as(u32, default_block_size), hdr.block_size);
    try testing.expect(hdr.content_size == null);
    try testing.expectEqual(n, hdr.header_size);
}

test "writeHeader / parseHeader roundtrip, with content_size" {
    var buf: [max_header_size]u8 = undefined;
    const n = try writeHeader(&buf, .{
        .codec = .high,
        .level = 9,
        .block_size = 1 << 18,
        .content_size = 1234567,
    });
    const hdr = try parseHeader(buf[0..n]);
    try testing.expectEqual(Codec.high, hdr.codec);
    try testing.expectEqual(@as(u8, 9), hdr.level);
    try testing.expectEqual(@as(?u64, 1234567), hdr.content_size);
    try testing.expect(hdr.flags.content_size_present);
}

test "parses real C# L1 fixture" {
    // First 18 bytes of c:/tmp/test_tiny_l1.slz observed earlier:
    //   31 5a 4c 53  01 01 01 01 02 00  29 00 00 00 00 00 00 00
    const fixture = [_]u8{
        0x31, 0x5A, 0x4C, 0x53, // magic 'SLZ1'
        0x01, // version
        0x01, // flags: content_size_present
        0x01, // codec: Fast
        0x01, // level: 1
        0x02, // blockSizeLog2 (−16) → 18 → 256 KB
        0x00, // reserved
        0x29, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // content_size = 41
    };
    const hdr = try parseHeader(&fixture);
    try testing.expectEqual(Codec.fast, hdr.codec);
    try testing.expectEqual(@as(u8, 1), hdr.level);
    try testing.expectEqual(@as(u32, 1 << 18), hdr.block_size);
    try testing.expectEqual(@as(?u64, 41), hdr.content_size);
    try testing.expectEqual(@as(usize, 18), hdr.header_size);
}

test "block header end mark detected" {
    const em = [_]u8{ 0, 0, 0, 0, 0, 0, 0, 0 };
    const bh = try parseBlockHeader(&em);
    try testing.expect(bh.isEndMark());
}

test "block header uncompressed flag roundtrip" {
    var buf: [8]u8 = undefined;
    writeBlockHeader(&buf, .{
        .compressed_size = 41,
        .decompressed_size = 41,
        .uncompressed = true,
    });
    const bh = try parseBlockHeader(&buf);
    try testing.expect(bh.uncompressed);
    try testing.expectEqual(@as(u32, 41), bh.compressed_size);
    try testing.expectEqual(@as(u32, 41), bh.decompressed_size);
}
